### ISSUE TITLE
The Ontology import lets Hets use http instead of https

### ISSUE THREAD

**Issue by eugenk** [ontohub/ontohub#1658]
When parsing an ontology, the URL that Hets shall fetch the ontology from is using HTTP instead of HTTPS, e.g. http://ontohub.org/ref/17/meta/dol-ontology.

This is because Rails is executed by puma, which only uses HTTP because Apache works as a reverse proxy, managing everything with TLS and puma doesn't know that the original request came in via HTTPS.

We need a way to find out whether or not we got the original request with TLS. @jelmd: Is there a way for Apache to tell Puma the original request URL?

**Comment by jelmd** [comment]
Actually there is no need to check, because since #1613 everything what puma sees, should be HTTPS based.

**Comment by eugenk** [comment]
Actually, puma only sees HTTP. I temporarily added some debug log printing:

```text
request.scheme: http
request.base_url: http://ontohub.org
request.original_url: http://ontohub.org/
request.session_options[:secure]: false
request.session_options[:httponly]: true
```

**Comment by eugenk** [comment]
For documentation of the methods called, see, for instance, http://api.rubyonrails.org/classes/ActionDispatch/Request.html#method-i-original_url

We don't have any indication whether or not TLS was used and cannot redirect to the URL scheme that we received the request with.

It's also very odd that Hets can't parse http://ontohub.org/ref/17/meta/dol-ontology, but it can parse https://ontohub.org/ref/17/meta/dol-ontology. You can try it with
```text
hets-server -X & # no need to run this on t[a-e]
curl http://localhost:8000/dg/http%3a%2f%2fontohub%2eorg%2fref%2f17%2fmeta%2fdol%2dontology/auto
curl http://localhost:8000/dg/https%3a%2f%2fontohub%2eorg%2fref%2f17%2fmeta%2fdol%2dontology/auto
```
If you receive XML output, everything is fine.

**Comment by jelmd** [comment]
1) Again, since #1613 clients are forced to use HTTPS, so there is no need to double check. At the same time HTTPs is forced for everything one can just use 'https' as _the_ scheme for redirection. Actually one could even say, it doesn't matter, whether the backend uses `http` or `https` to generate URLs because any HTTP request get turned into HTTPS anyway. Last but not least my experience has shown, that it is not very wise to generate full URLs (i.e. incl. scheme, host, port, user, passwd) unless one really wants to refer to a different site. This always lead sooner or later to problems ...

2) Why do you care about TLS. Do you want to implement/fake the security layer as well? Simple advice: don't!!!

3) Because of `ProxyPass / http://localhost:3000/` it seems to be clear, why it sees HTTP only. Not sure, whether puma would understand HTTPs as well (since IIRC there are no knobs, to set CA bundle, CRT/CRLs to use what a server obviously needs), but even if, it would make no sense at all to decrypt and encrypt all the stuff twice on the same zone.

4) If an app (the backend) works different/produces different results depending on whether HTTP or HTTPS, this sounds like a bug, yepp.

**Comment by eugenk** [comment]
To 1) Well, it sometimes refers to hosts other than `localhost`. Hets runs on different machines and all of them access the ontohub.org host. Hence we need to generate full URLs.
I understand that 

2) I only care about correctly resolved URLs and mentioned TLS because it is part of HTTPS

3) Is there a way for puma to know that the original request was sent with the https scheme? That's the most important question of this issue.

4) True, but I can't find the bug itself. Hets retrieves the file with the system call
```
wget --header="Accept: */*; q=0.1, text/plain" --no-check-certificate -O - http://ontohub.org/ref/17/meta/dol-ontology
```
and this gives us the correct file even for HTTP...

**Comment by jelmd** [comment]
1..4-3) ok
3) Probably via custom header injection or something like this. Checking on td ...

**Comment by jelmd** [comment]
3) The apache httpd now injects a `X-Scheme:` header into each request. The value is the SSL protocol used (so for HTTPS something like `TLSv1.2`, for HTTP `(null)`).

4) We (till+me) tried to find an explaination for it. Initial idea was, that it might be caused by redirection, but doesn't make much sense on the other side, because redirection is handled automagically by wget/curl. However, `hets http://ontohub.org/ref/17/meta/dol-ontology` once with https and once with http produce different results, yepp. Unfortunately `curl --noproxy ontohub.org --header 'Accept: */*; q=0.1, text/plain' --insecure -s -S http://ontohub.org/ref/17/meta/dol-ontology` once with https and once with http produce exactly the same output (on Solaris hets uses curl instead of wget). Strange ...

**Comment by jelmd** [comment]
Well, I patched the owlapi to dump the document to parse on demand, and yes, it tries to parse the redirect response. Because the called curl reads my ~/.curlrc, redirection seems to be used and work as shown above. So the only explanation is, that somewhere else the doc gets fetched again and than redirection gets not handled properly.

Access logs seem to agree:
```
T1: 301 "GET /ref/17/meta/dol-ontology HTTP/1.1" "-" "curl/7.45.0" 648       # the redirect response
T2: 200 "GET /ref/17/meta/dol-ontology HTTP/1.1" "-" "curl/7.45.0" 173699 # the RDF document
# again?
T3: 301 "GET /ref/17/meta/dol-ontology HTTP/1.1" "-" "curl/7.45.0" 648       # the redirect response
T4: 200 "GET /ref/17/meta/dol-ontology HTTP/1.1" "-" "curl/7.45.0" 173699 # the RDF document
# and
T5: 301 "GET /ref/17/meta/dol-ontology HTTP/1.1" "-" "Java/1.8.0_72" 648  # the redirect response
```
So very odd - one `hets http://ontohub.org/ref/17/meta/dol-ontology` cause it to get pulled 3 times, and the 3rd time it fails because of not handling the redirect response properly. With https it pulls it 3x times as well, but it works, because there is finally no redirect ...

PS: If you wanna try the dump by yourself: http://iws.cs.uni-magdeburg.de/~elkner/ontohub/ contains the jar to replace in lib/hets/hets-owl-tools/lib/ . Set  `TMPDOC_NAME` to e.g. /tmp/out and start hets ...

**Comment by tillmo** [comment]
Since this is primarily a Hets issue, I have created https://github.com/spechub/Hets/issues/1620

**Comment by tillmo** [comment]
solved by spechub/Hets#1620
